src/tactic/itauto.lean

```diff
@@ -0,0 +1,486 @@
+| _ A B := (A.imp B, B.imp A)
+
+/-- Debugging printer for propositions. -/
+meta def prop.to_format : prop → format
+| (prop.var i) := format!"v{i}"
+| prop.true := format!"⊤"
+| prop.false := format!"⊥"
+| (prop.and and_kind.and p q) := format!"({p.to_format} ∧ {q.to_format})"
+| (prop.and and_kind.iff p q) := format!"({p.to_format} ↔ {q.to_format})"
+| (prop.and and_kind.eq p q) := format!"({p.to_format} = {q.to_format})"
+| (prop.or p q) := format!"({p.to_format} ∨ {q.to_format})"
+| (prop.imp p q) := format!"({p.to_format} → {q.to_format})"
+
+meta instance : has_to_format prop := ⟨prop.to_format⟩
+
+section
+open ordering
+
+/-- A comparator for `and_kind`. (There should really be a derive handler for this.) -/
+def and_kind.cmp (p q : and_kind) : ordering :=
+by { cases p; cases q, exacts [eq, lt, lt, gt, eq, lt, gt, gt, eq] }
+
+/-- A comparator for propositions. (There should really be a derive handler for this.) -/
+def prop.cmp (p q : prop) : ordering :=
+begin
+  induction p with _ ap _ _ p₁ p₂ _ _ p₁ p₂ _ _ p₁ p₂ _ _ p₁ p₂ generalizing q; cases q,
+  case var var { exact cmp p q },
+  case true true { exact eq },
+  case false false { exact eq },
+  case and and : aq q₁ q₂ { exact (ap.cmp aq).or_else ((p₁ q₁).or_else (p₂ q₂)) },
+  case or or : q₁ q₂ { exact (p₁ q₁).or_else (p₂ q₂) },
+  case imp imp : q₁ q₂ { exact (p₁ q₁).or_else (p₂ q₂) },
+  exacts [lt, lt, lt, lt, lt,
+          gt, lt, lt, lt, lt,
+          gt, gt, lt, lt, lt,
+          gt, gt, gt, lt, lt,
+          gt, gt, gt, gt, lt,
+          gt, gt, gt, gt, gt]
+end
+
+instance : has_lt prop := ⟨λ p q, p.cmp q = lt⟩
+
+instance : decidable_rel (@has_lt.lt prop _) := λ _ _, ordering.decidable_eq _ _
+
+end
+
+/-- A reified inductive proof type for intuitionistic propositional logic. -/
+@[derive has_reflect]
+inductive proof
+-- (n: A) ⊢ A
+| hyp (n : name) : proof
+-- ⊢ ⊤
+| triv : proof
+-- (p: ⊥) ⊢ A
+| exfalso' (A : prop) (p : proof) : proof
+-- (p: (x: A) ⊢ B) ⊢ A → B
+| intro (x : name) (A : prop) (p : proof) : proof
+-- ak = and:  (p: A ∧ B) ⊢ A
+-- ak = iff:  (p: A ↔ B) ⊢ A → B
+-- ak = eq:  (p: A = B) ⊢ A → B
+| and_left (ak : and_kind) (p : proof) : proof
+-- ak = and:  (p: A ∧ B) ⊢ B
+-- ak = iff:  (p: A ↔ B) ⊢ B → A
+-- ak = eq:  (p: A = B) ⊢ B → A
+| and_right (ak : and_kind) (p : proof) : proof
+-- ak = and:  (p₁: A) (p₂: B) ⊢ A ∧ B
+-- ak = iff:  (p₁: A → B) (p₁: B → A) ⊢ A ↔ B
+-- ak = eq:  (p₁: A → B) (p₁: B → A) ⊢ A = B
+| and_intro (ak : and_kind) (p₁ p₂ : proof) : proof
+-- ak = and:  (p: A ∧ B → C) ⊢ A → B → C
+-- ak = iff:  (p: (A ↔ B) → C) ⊢ (A → B) → (B → A) → C
+-- ak = eq:  (p: (A = B) → C) ⊢ (A → B) → (B → A) → C
+| curry (ak : and_kind) (A B : prop) (p : proof) : proof
+-- This is a partial application of curry.
+-- ak = and:  (p: A ∧ B → C) (q : A) ⊢ B → C
+-- ak = iff:  (p: (A ↔ B) → C) (q: A → B) ⊢ (B → A) → C
+-- ak = eq:  (p: (A ↔ B) → C) (q: A → B) ⊢ (B → A) → C
+| curry₂ (ak : and_kind) (B : prop) (p q : proof) : proof
```

Comment:
`curry2` isn't used directly by the proof generator, it only produces `curry`. This is an intermediate term created by `app`, which does a bit of opportunistic proof simplification. The alternative would be to do the proof simplification in `apply_proof`, by directly pattern matching on `app' (curry ...) _` and `app' (app' (curry ...) _) _`, but I think this would slow it down because of the large pattern match.